### PR TITLE
FIX: Show a correct diff when editing consecutive paragraphs

### PR DESCRIPTION
**Pull request by nachocab**
Currently, if two consecutive paragraphs are edited, their contents are not properly diffed:
Screen Shot 2019-10-10 at 09 35 13

The current algorithm performs a pairwise diff whenever a deletion follows an addition or vice versa. This happens here:

https://github.com/discourse/discourse/blob/c5326682d6346c13b495ece9f3ee4e82f1b13a72/lib/discourse_diff.rb#L38-L39

However, in the example above we have two additions followed by two deletions. This leads to:
- an added paragraph: `this is one great paragraph`
- a pairwise comparison: `this is one paragraph` vs. `this is another` 
- a deleted paragraph: `here is yet another`

The ideal approach would be to perform an alignment step before diffing (which is not trivial to implement, see [Gale-Church algorithm](https://en.wikipedia.org/wiki/Gale–Church_alignment_algorithm)), so a good compromise is to use one of the alternative edit sequences to the one returned by ONPDiff. 

In other words, `ONPDiff.paragraph_diff` transforms the edit sequence from `ONPDiff.diff` `Add1 Add2 Delete1 Delete2` into this one `Add1 Delete1 Add2 Delete2`. 

Current behavior using `ONPDiff.diff`:

image

Proposed behavior using `ONPDiff.paragraph_diff`:

image

### CODE REVIEW

**Comment by ZogStriP** [comment]
Wow 😍

SO MUCH BETTERER 👏

**Comment by ZogStriP** [comment]
Can you benchmark this on very large diffs? Both when there are lots of text but a small number of diffs or a large-ish amount of text but a lot of diffs.

**Comment by nachocab** [comment]
> Can you benchmark this on very large diffs? Both when there are lots of text but a small number of diffs or a large-ish amount of text but a lot of diffs.
Sure, do you need me to use a specific tool or just to sanity check it on my computer?

**Comment by ZogStriP** [comment]
> Sure, do you need me to use a specific tool or just to sanity check it on my computer?

Sanity check is fine as long as I have before/after ;)

> And how big do you want the diffs?

Regarind text size, near the default maximum post lenght.

**Comment by nachocab** [comment]
They were both equally fast (less than a second, probably less than .5s): 

image
